payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java

```diff
@@ -398,6 +398,8 @@ public boolean apply(final PaymentTransactionModelDao input) {
                     // We can't tell where we should be in the state machine - bail (cannot be enforced by the state machine unfortunately because UNKNOWN and PLUGIN_FAILURE are both treated as EXCEPTION)
                     if (transactionToComplete.getTransactionStatus() == TransactionStatus.UNKNOWN) {
                         throw new PaymentApiException(ErrorCode.PAYMENT_INVALID_OPERATION, paymentStateContext.getTransactionType(), transactionToComplete.getTransactionStatus());
+                    } else if (transactionToComplete.getTransactionStatus() == TransactionStatus.SUCCESS ){
+                        throw new PaymentApiException(ErrorCode.PAYMENT_ACTIVE_TRANSACTION_KEY_EXISTS, paymentStateContext.getPaymentTransactionExternalKey());
```

Comment:
> Suggested fix instead:
Move invokeJanitor to line 388, i.e. if there is a transaction id or key, always invoke the Janitor to fix the state

It is there currently, I think you mean put it before the original Sanity checks? But, the `findTransactionToCompleteAndRunSanityChecks` method returned the transaction to complete. Can't invoke Janitor before it.

>Move runSanityOnTransactionExternalKey to line 412, i.e. after having run the Janitor and updated the payment state context

I think this is better and doable, but need more extra and duplicated code to extract the transaction from the collection if put the sanity check here.

----
So I think the current fixing is a better solution. How do you think? @pierre